File: modules/realtime-recorder/ios/OverlayRenderer.swift
import UIKit
import CoreGraphics
import CoreMedia
import CoreText

/// Holds the current overlay state pushed from JS
struct OverlayState {
  var timerType: String = ""
  var timerDisplay: String = ""       // e.g. "02:35.42"
  var title: String = ""
  var timestamp: String = ""
  var isRecording: Bool = false
  var countdownValue: Int = 0         // >0 means countdown is visible
  var showTimer: Bool = false         // true when chrono is running/frozen
  var boxLogoUrl: String = ""         // URL of the box logo (empty = no box)
  var competitionLogoUrl: String = "" // URL of competition logo (top-left overlay)
}

/// Draws overlay graphics directly onto a CVPixelBuffer using Core Graphics
final class OverlayRenderer {

  // Cached images
  private var cachedAthlexLogo: UIImage?
  private var cachedBoxLogo: UIImage?
  private var cachedBoxLogoUrl: String = ""
  private var boxLogoLoading = false
  private var cachedCompLogo: UIImage?
  private var cachedCompLogoUrl: String = ""
  private var compLogoLoading = false
  private var blackOpsFont: UIFont?
  private var dsDigitalFont: UIFont?

  init() {
    loadAthlexLogo()
    loadBlackOpsFont()
    loadDSDigitalFont()
  }

  // MARK: - Logo loading

  private func loadAthlexLogo() {
    // Try the resource bundle first (CocoaPods resource_bundles)
    let bundleName = "RealtimeRecorderResources"
    if let bundleURL = Bundle.main.url(forResource: bundleName, withExtension: "bundle"),
       let resBundle = Bundle(url: bundleURL),
       let img = UIImage(named: "logo", in: resBundle, compatibleWith: nil) ?? UIImage(contentsOfFile: resBundle.path(forResource: "logo", ofType: "png") ?? "") {
      cachedAthlexLogo = img
    }
    // Fallback: main bundle
    else if let img = UIImage(named: "logo") ?? UIImage(named: "logo.png") {
      cachedAthlexLogo = img
    }
  }

  private func loadBlackOpsFont() {
    let bundleName = "RealtimeRecorderResources"
    guard let bundleURL = Bundle.main.url(forResource: bundleName, withExtension: "bundle"),
          let resBundle = Bundle(url: bundleURL),
          let fontURL = resBundle.url(forResource: "BlackOpsOne", withExtension: "ttf") else {
      print("[OverlayRenderer] BlackOpsOne.ttf not found in resource bundle")
      return
    }
    var errorRef: Unmanaged<CFError>?
    CTFontManagerRegisterFontsForURL(fontURL as CFURL, .process, &errorRef)
    if let err = errorRef?.takeRetainedValue() {
      // Already registered is OK
      let desc = CFErrorGetDomain(err) as String
      if !desc.contains("already registered") {
        print("[OverlayRenderer] Font registration error: \(err)")
      }
    }
    // PostScript name for Black Ops One is "BlackOpsOne-Regular"
    if let font = UIFont(name: "BlackOpsOne-Regular", size: 48) {
      blackOpsFont = font
    } else {
      print("[OverlayRenderer] BlackOpsOne-Regular font not available after registration")
    }
  }

  private func loadDSDigitalFont() {
    let bundleName = "RealtimeRecorderResources"
    guard let bundleURL = Bundle.main.url(forResource: bundleName, withExtension: "bundle"),
          let resBundle = Bundle(url: bundleURL),
          let fontURL = resBundle.url(forResource: "DS-Digital", withExtension: "ttf") else {
      print("[OverlayRenderer] DS-Digital.ttf not found in resource bundle")
      return
    }
    var errorRef: Unmanaged<CFError>?
    CTFontManagerRegisterFontsForURL(fontURL as CFURL, .process, &errorRef)
    if let err = errorRef?.takeRetainedValue() {
      let desc = CFErrorGetDomain(err) as String
      if !desc.contains("already registered") {
        print("[OverlayRenderer] DS-Digital font registration error: \(err)")
      }
    }
    // PostScript name for DS-Digital is "DS-Digital"
    if let font = UIFont(name: "DS-Digital", size: 48) {
      dsDigitalFont = font
    } else {
      print("[OverlayRenderer] DS-Digital font not available after registration")
    }
  }

  private func loadBoxLogoIfNeeded(url: String) {
    guard !url.isEmpty, url != cachedBoxLogoUrl, !boxLogoLoading else { return }
    boxLogoLoading = true
    cachedBoxLogoUrl = url

    DispatchQueue.global(qos: .utility).async { [weak self] in
      guard let self = self, let imgURL = URL(string: url),
            let data = try? Data(contentsOf: imgURL),
            let img = UIImage(data: data) else {
        self?.boxLogoLoading = false
        return
      }
      self.cachedBoxLogo = img
      self.boxLogoLoading = false
    }
  }

  private func loadCompLogoIfNeeded(url: String) {
    guard !url.isEmpty, url != cachedCompLogoUrl, !compLogoLoading else { return }
    compLogoLoading = true
    cachedCompLogoUrl = url

    DispatchQueue.global(qos: .utility).async { [weak self] in
      guard let self = self, let imgURL = URL(string: url),
            let data = try? Data(contentsOf: imgURL),
            let img = UIImage(data: data) else {
        self?.compLogoLoading = false
        return
      }
      self.cachedCompLogo = img
      self.compLogoLoading = false
    }
  }

  // MARK: - Main render

  /// Draw all overlays onto the given pixel buffer. Thread-safe — called from capture queue.
  func render(onto pixelBuffer: CVPixelBuffer, state: OverlayState) {
    let width = CVPixelBufferGetWidth(pixelBuffer)
    let height = CVPixelBufferGetHeight(pixelBuffer)
    let size = CGSize(width: CGFloat(width), height: CGFloat(height))
    let isLandscape = size.width > size.height

    CVPixelBufferLockBaseAddress(pixelBuffer, [])
    defer { CVPixelBufferUnlockBaseAddress(pixelBuffer, []) }

    guard let context = CGContext(
      data: CVPixelBufferGetBaseAddress(pixelBuffer),
      width: width,
      height: height,
      bitsPerComponent: 8,
      bytesPerRow: CVPixelBufferGetBytesPerRow(pixelBuffer),
      space: CGColorSpaceCreateDeviceRGB(),
      bitmapInfo: CGBitmapInfo.byteOrder32Little.rawValue | CGImageAlphaInfo.premultipliedFirst.rawValue
    ) else { return }

    // Core Graphics origin is bottom-left; UIKit is top-left.
    context.translateBy(x: 0, y: size.height)
    context.scaleBy(x: 1, y: -1)

    // Load logos in background if URL changed
    loadBoxLogoIfNeeded(url: state.boxLogoUrl)
    loadCompLogoIfNeeded(url: state.competitionLogoUrl)

    // Use min dimension as reference so elements stay the same physical size
    let refDim = min(size.width, size.height)
    let scale = refDim / 1080.0
    let margin: CGFloat = 24 * scale
    let safeTop: CGFloat = isLandscape ? 24 * scale : 60 * (size.height / 1920.0)

    // ─── 0. Competition logo (top left — rounded square, no white bg) ───
    let logoSize: CGFloat = isLandscape ? 120 * scale : 200 * scale
    if let compImg = cachedCompLogo {
      let logoRect = CGRect(x: margin, y: safeTop, width: logoSize, height: logoSize)
      let cornerRadius: CGFloat = isLandscape ? 20 * scale : 32 * scale
      UIGraphicsPushContext(context)
      context.saveGState()
      let path = UIBezierPath(roundedRect: logoRect, cornerRadius: cornerRadius)
      path.addClip()
      compImg.draw(in: logoRect)
      context.restoreGState()
      UIGraphicsPopContext()
    }

    // ─── 1. Title (top center) ───
    if !state.title.isEmpty {
      let titleX = cachedCompLogo != nil ? (margin + logoSize + 12 * scale) : margin
      let titleW = size.width - titleX - (cachedBoxLogo != nil ? (logoSize + margin + 12 * scale) : margin)
      drawText(context: context, text: state.title,
               rect: CGRect(x: titleX, y: safeTop, width: titleW, height: 40 * scale),
               fontSize: 28 * scale, bold: true, color: .white, alignment: .center, shadow: true)
    }

    // ─── 2. Box logo (top right — circle, no background) ───
    if let boxImg = cachedBoxLogo {
      let logoRect = CGRect(x: size.width - logoSize - margin, y: safeTop, width: logoSize, height: logoSize)
      let cornerRadius: CGFloat = logoSize / 2  // circle
      UIGraphicsPushContext(context)
      context.saveGState()
      let path = UIBezierPath(roundedRect: logoRect, cornerRadius: cornerRadius)
      path.addClip()
      boxImg.draw(in: logoRect)
      context.restoreGState()
      UIGraphicsPopContext()
    }

    // ─── 3. Countdown (center, extra large, bold) ───
    if state.countdownValue > 0 {
      let cdStr = "\(state.countdownValue)"
      let cdFontSize: CGFloat = isLandscape ? 180 * scale : 260 * scale
      let cdH: CGFloat = isLandscape ? 220 * scale : 320 * scale
      drawText(context: context, text: cdStr,
               rect: CGRect(x: 0, y: (size.height - cdH) / 2, width: size.width, height: cdH),
               fontSize: cdFontSize, bold: true, color: .white, alignment: .center, weight: .bold)
    }

    // ════════════════════════════════════════════
    //  BOTTOM ROW — AthleX (left) | Timer (center) | Timestamp (right)
    //  Logo centered above "AthleX" text
    // ════════════════════════════════════════════
    let safeBottom: CGFloat = isLandscape ? 40 * scale : 140 * scale

    // ─── 6. "AthleX" branded text — reference baseline for bottom row ───
    let brandH: CGFloat = isLandscape ? 40 * scale : 50 * scale
    let brandY = size.height - safeBottom - brandH
    let brandFontSize: CGFloat = isLandscape ? 30 * scale : 40 * scale
    let brandTextW: CGFloat = isLandscape ? 120 * scale : 160 * scale
    drawBrandText(context: context,
                  rect: CGRect(x: margin, y: brandY, width: 280 * scale, height: brandH),
                  fontSize: brandFontSize, color: .white, shadow: true)

    // ─── 5. ATHLEX logo (centered horizontally above "AthleX" text) ───
    let atlLogoH: CGFloat = isLandscape ? 80 * scale : 120 * scale
    let atlLogoY = brandY - atlLogoH - 4 * scale
    if let atlImg = cachedAthlexLogo {
      let atlLogoW = atlLogoH * (atlImg.size.width / atlImg.size.height)
      let brandCenterX = margin + brandTextW / 2
      let logoX = brandCenterX - atlLogoW / 2
      let logoRect = CGRect(x: logoX, y: atlLogoY, width: atlLogoW, height: atlLogoH)
      UIGraphicsPushContext(context)
      atlImg.draw(in: logoRect)
      UIGraphicsPopContext()
    }

    // ─── 4. Timer display (center, same row as AthleX text) ───
    if state.showTimer && state.countdownValue <= 0 {
      let timerFontSize: CGFloat = isLandscape ? 70 * scale : 90 * scale
      let timerH: CGFloat = isLandscape ? 85 * scale : 110 * scale
      let timerCenterY = brandY + brandH / 2
      let timerY = timerCenterY - timerH / 2
      drawText(context: context, text: state.timerDisplay,
               rect: CGRect(x: 0, y: timerY, width: size.width, height: timerH),
               fontSize: timerFontSize, bold: false, color: .white, alignment: .center,
               weight: .medium, shadow: true, dsDigital: true)
    }

    // ─── 7. Timestamp (right, same row as AthleX text) ───
    if !state.timestamp.isEmpty && state.showTimer && state.countdownValue <= 0 {
      let tsW: CGFloat = 260 * scale
      let tsH: CGFloat = 34 * scale
      let tsCenterY = brandY + brandH / 2
      let tsY = tsCenterY - tsH / 2
      drawText(context: context, text: state.timestamp,
               rect: CGRect(x: size.width - tsW - margin, y: tsY, width: tsW, height: tsH),
               fontSize: 24 * scale, bold: false, color: UIColor.white.withAlphaComponent(0.8),
               alignment: .right, shadow: true)
    }
  }

  // MARK: - Brand text (Black Ops One)

  private func drawBrandText(
    context: CGContext,
    rect: CGRect,
    fontSize: CGFloat,
    color: UIColor,
    shadow: Bool
  ) {
    let font: UIFont = blackOpsFont?.withSize(fontSize)
      ?? UIFont.boldSystemFont(ofSize: fontSize) // fallback

    let paragraphStyle = NSMutableParagraphStyle()
    paragraphStyle.alignment = .left
    paragraphStyle.lineBreakMode = .byClipping

    var attributes: [NSAttributedString.Key: Any] = [
      .font: font,
      .foregroundColor: color,
      .paragraphStyle: paragraphStyle,
    ]

    if shadow {
      let s = NSShadow()
      s.shadowColor = UIColor.black.withAlphaComponent(0.7)
      s.shadowOffset = CGSize(width: 2, height: 2)
      s.shadowBlurRadius = 6
      attributes[.shadow] = s
    }

    let attrString = NSAttributedString(string: "AthleX", attributes: attributes)

    UIGraphicsPushContext(context)
    attrString.draw(in: rect)
    UIGraphicsPopContext()
  }

  // MARK: - Text drawing helper

  private func drawText(
    context: CGContext,
    text: String,
    rect: CGRect,
    fontSize: CGFloat,
    bold: Bool,
    color: UIColor,
    alignment: NSTextAlignment,
    weight: UIFont.Weight? = nil,
    shadow: Bool = false,
    monospace: Bool = false,
    dsDigital: Bool = false
  ) {
    let font: UIFont
    if dsDigital, let dsFont = dsDigitalFont?.withSize(fontSize) {
      font = dsFont
    } else if monospace {
      font = UIFont.monospacedDigitSystemFont(ofSize: fontSize, weight: weight ?? (bold ? .bold : .regular))
    } else if let w = weight {
      font = UIFont.systemFont(ofSize: fontSize, weight: w)
    } else {
      font = bold ? UIFont.boldSystemFont(ofSize: fontSize) : UIFont.systemFont(ofSize: fontSize)
    }

    let paragraphStyle = NSMutableParagraphStyle()
    paragraphStyle.alignment = alignment
    paragraphStyle.lineBreakMode = .byTruncatingTail

    var attributes: [NSAttributedString.Key: Any] = [
      .font: font,
      .foregroundColor: color,
      .paragraphStyle: paragraphStyle,
    ]

    if shadow {
      let s = NSShadow()
      s.shadowColor = UIColor.black.withAlphaComponent(0.7)
      s.shadowOffset = CGSize(width: 1, height: 1)
      s.shadowBlurRadius = 4
      attributes[.shadow] = s
    }

    let attrString = NSAttributedString(string: text, attributes: attributes)

    UIGraphicsPushContext(context)
    attrString.draw(in: rect)
    UIGraphicsPopContext()
  }
}

File: modules/realtime-recorder/android/src/main/java/expo/modules/realtimerecorder/OverlayRenderer.kt
package expo.modules.realtimerecorder

import android.content.Context
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.text.Layout
import android.text.StaticLayout
import android.text.TextPaint
import java.net.URL

/**
 * Draws overlay graphics directly onto a Bitmap (video frame) using Android Canvas.
 * Mirrors the iOS OverlayRenderer.swift layout exactly:
 *
 * - Competition logo: top left (rounded square)
 * - Title: top center (28pt bold)
 * - Box logo: top right (circle)
 * - Countdown: center screen (260pt bold)
 * - Bottom row (all vertically centered):
 *   - Left: ATHLEX logo (160px)
 *   - Center: Timer DS-Digital (180pt)
 *   - Right: Timestamp
 */
class OverlayRenderer(private val context: Context) {

  @Volatile private var cachedAthlexLogo: Bitmap? = null
  @Volatile private var cachedBoxLogo: Bitmap? = null
  @Volatile private var cachedBoxLogoUrl: String = ""
  @Volatile private var boxLogoLoading = false
  @Volatile private var cachedCompLogo: Bitmap? = null
  @Volatile private var cachedCompLogoUrl: String = ""
  @Volatile private var compLogoLoading = false
  private var dsDigitalTypeface: Typeface? = null

  // Pre-allocated objects to avoid GC pressure on every frame
  private val textPaint = TextPaint(Paint.ANTI_ALIAS_FLAG)
  private val bitmapPaint = Paint(Paint.FILTER_BITMAP_FLAG)
  private val reusableClipPath = Path()

  init {
    loadAthlexLogo()
    loadDSDigitalFont()
  }

  // MARK: - Resource loading

  private fun loadAthlexLogo() {
    try {
      val inputStream = context.assets.open("realtime-recorder/logo.png")
      cachedAthlexLogo = BitmapFactory.decodeStream(inputStream)
      inputStream.close()
    } catch (e: Exception) {
      // Try raw resources
      try {
        val resId = context.resources.getIdentifier("logo", "drawable", context.packageName)
        if (resId != 0) {
          cachedAthlexLogo = BitmapFactory.decodeResource(context.resources, resId)
        }
      } catch (_: Exception) {}
    }
  }

  private fun loadDSDigitalFont() {
    try {
      dsDigitalTypeface = Typeface.createFromAsset(context.assets, "realtime-recorder/DS-Digital.ttf")
    } catch (e: Exception) {
      dsDigitalTypeface = Typeface.create(Typeface.MONOSPACE, Typeface.NORMAL)
    }
  }

  private fun loadBoxLogoIfNeeded(url: String) {
    if (url.isEmpty() || url == cachedBoxLogoUrl || boxLogoLoading) return
    boxLogoLoading = true
    cachedBoxLogoUrl = url

    Thread {
      try {
        val stream = URL(url).openStream()
        val bmp = BitmapFactory.decodeStream(stream)
        stream.close()
        cachedBoxLogo = bmp
      } catch (_: Exception) {
      } finally {
        boxLogoLoading = false
      }
    }.start()
  }

  private fun loadCompLogoIfNeeded(url: String) {
    if (url.isEmpty() || url == cachedCompLogoUrl || compLogoLoading) return
    compLogoLoading = true
    cachedCompLogoUrl = url

    Thread {
      try {
        val stream = URL(url).openStream()
        val bmp = BitmapFactory.decodeStream(stream)
        stream.close()
        cachedCompLogo = bmp
      } catch (_: Exception) {
      } finally {
        compLogoLoading = false
      }
    }.start()
  }

  // MARK: - Main render

  /**
   * Draw all overlays onto the given Bitmap. Called from recording thread.
   */
  fun render(bitmap: Bitmap, state: OverlayState) {
    val canvas = Canvas(bitmap)
    val width = bitmap.width.toFloat()
    val height = bitmap.height.toFloat()
    val isLandscape = width > height

    loadBoxLogoIfNeeded(state.boxLogoUrl)
    loadCompLogoIfNeeded(state.competitionLogoUrl)

    // Use min dimension as reference so elements stay the same physical size
    val refDim = minOf(width, height)
    val scale = refDim / 1080f
    val margin = 24f * scale
    val safeTop = if (isLandscape) 24f * scale else 60f * (height / 1920f)

    // ─── 0. Competition logo (top left — rounded square, no white bg) ───
    val logoSize = if (isLandscape) 120f * scale else 200f * scale
    cachedCompLogo?.let { compImg ->
      val logoRect = RectF(margin, safeTop, margin + logoSize, safeTop + logoSize)
      val cornerRadius = if (isLandscape) 20f * scale else 32f * scale

      canvas.save()
      reusableClipPath.reset()
      reusableClipPath.addRoundRect(logoRect, cornerRadius, cornerRadius, Path.Direction.CW)
      canvas.clipPath(reusableClipPath)
      canvas.drawBitmap(compImg, null, logoRect, bitmapPaint)
      canvas.restore()
    }

    // ─── 1. Title (top center, adjusted for logos) ───
    if (state.title.isNotEmpty()) {
      val titleLeft = if (cachedCompLogo != null) margin + logoSize + 12f * scale else margin
      val titleRight = if (cachedBoxLogo != null) width - logoSize - margin - 12f * scale else width - margin
      drawText(
        canvas, state.title,
        RectF(titleLeft, safeTop, titleRight, safeTop + 40f * scale),
        fontSize = 28f * scale, bold = true, color = Color.WHITE,
        alignment = Layout.Alignment.ALIGN_CENTER, shadow = true
      )
    }

    // ─── 2. Box logo (top right — circle, no background) ───
    cachedBoxLogo?.let { boxImg ->
      val logoRect = RectF(
        width - logoSize - margin, safeTop,
        width - margin, safeTop + logoSize
      )
      val cornerRadius = logoSize / 2f  // circle

      canvas.save()
      reusableClipPath.reset()
      reusableClipPath.addRoundRect(logoRect, cornerRadius, cornerRadius, Path.Direction.CW)
      canvas.clipPath(reusableClipPath)
      canvas.drawBitmap(boxImg, null, logoRect, bitmapPaint)
      canvas.restore()
    }

    // ─── 3. Countdown (center, extra large, bold) ───
    if (state.countdownValue > 0) {
      val cdStr = "${state.countdownValue}"
      val cdFontSize = if (isLandscape) 180f * scale else 260f * scale
      val cdH = if (isLandscape) 220f * scale else 320f * scale
      val cdY = (height - cdH) / 2f
      drawText(
        canvas, cdStr,
        RectF(0f, cdY, width, cdY + cdH),
        fontSize = cdFontSize, bold = true, color = Color.WHITE,
        alignment = Layout.Alignment.ALIGN_CENTER
      )
    }

    // ════════════════════════════════════════════
    //  BOTTOM ROW — AthleX logo (left) | Timer (center) | Timestamp (right)
    //  All elements vertically centered on the same row
    // ════════════════════════════════════════════
    val safeBottom = if (isLandscape) 40f * scale else 140f * scale

    // Row height driven by the timer (largest element)
    val timerFontSize = if (isLandscape) 140f * scale else 180f * scale
    val timerH = if (isLandscape) 170f * scale else 220f * scale
    val rowCenterY = height - safeBottom - timerH / 2f

    // ─── 4. AthleX logo (bottom-left, vertically centered) ───
    val atlLogoH = if (isLandscape) 120f * scale else 160f * scale
    cachedAthlexLogo?.let { atlImg ->
      val atlLogoW = atlLogoH * (atlImg.width.toFloat() / atlImg.height.toFloat())
      val logoY = rowCenterY - atlLogoH / 2f
      val logoRect = RectF(margin, logoY, margin + atlLogoW, logoY + atlLogoH)
      canvas.drawBitmap(atlImg, null, logoRect, bitmapPaint)
    }

    // ─── 5. Timer display (center, x2 size) ───
    if (state.showTimer && state.countdownValue <= 0) {
      val timerY = rowCenterY - timerH / 2f
      drawText(
        canvas, state.timerDisplay,
        RectF(0f, timerY, width, timerY + timerH),
        fontSize = timerFontSize, bold = false, color = Color.WHITE,
        alignment = Layout.Alignment.ALIGN_CENTER,
        shadow = true, dsDigital = true
      )
    }

    // ─── 6. Timestamp (right, vertically centered on same row) ───
    if (state.timestamp.isNotEmpty() && state.showTimer && state.countdownValue <= 0) {
      val tsW = 260f * scale
      val tsH = 34f * scale
      val tsY = rowCenterY - tsH / 2f
      drawText(
        canvas, state.timestamp,
        RectF(width - tsW - margin, tsY, width - margin, tsY + tsH),
        fontSize = 24f * scale, bold = false,
        color = Color.argb(204, 255, 255, 255),
        alignment = Layout.Alignment.ALIGN_OPPOSITE, shadow = true
      )
    }
  }

  // MARK: - Text drawing helper

  private fun drawText(
    canvas: Canvas, text: String, rect: RectF,
    fontSize: Float, bold: Boolean, color: Int,
    alignment: Layout.Alignment,
    shadow: Boolean = false,
    monospace: Boolean = false,
    dsDigital: Boolean = false
  ) {
    textPaint.reset()
    textPaint.isAntiAlias = true
    textPaint.color = color
    textPaint.textSize = fontSize
    textPaint.typeface = when {
      dsDigital -> dsDigitalTypeface ?: Typeface.create(Typeface.MONOSPACE, Typeface.NORMAL)
      monospace -> Typeface.create(Typeface.MONOSPACE, if (bold) Typeface.BOLD else Typeface.NORMAL)
      bold -> Typeface.DEFAULT_BOLD
      else -> Typeface.DEFAULT
    }
    if (shadow) {
      textPaint.setShadowLayer(4f, 1f, 1f, Color.argb(179, 0, 0, 0))
    }

    // Use StaticLayout for proper alignment
    val layoutWidth = (rect.right - rect.left).toInt().coerceAtLeast(1)

    @Suppress("DEPRECATION")
    val layout = StaticLayout(
      text, textPaint, layoutWidth,
      alignment, 1f, 0f, false
    )

    canvas.save()
    // Center vertically within rect
    val textHeight = layout.height.toFloat()
    val offsetY = rect.top + (rect.height() - textHeight) / 2f
    canvas.translate(rect.left, offsetY)
    layout.draw(canvas)
    canvas.restore()
  }
}
